### PR TITLE
feat(smart-contracts): make the library generic

### PR DESCRIPTION
**Pull request by fmeichtri-simtlix**
- Make contracts library generic to the chain
- Add Base classes
- Add Base interfaces
- Add Tron Structure
- Fix tests
- Update examples

Associated ticket [STLX-5](https://bitgoinc.atlassian.net/browse/STLX-5)
Reference to the Old deprecated PR: https://github.com/BitGo/eth-contracts/pull/19

### CODE REVIEW

File: src/base/methods/methods.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+  private parse(methods: M[]): BaseMethodContainerMap<M> {
+    return methods.reduce((collector: BaseMethodContainerMap<M>, method: M) => {
+      collector[method.name] = method;
+      return collector;
+    }, {});
+  }
+}
+
+export interface BaseMethodContainerMap<k extends Method> {
```
Comment (twtaylorbitgo) [inline src/base/methods/methods.ts:58]: let's try to capitalize our generics...

File: src/eth/contracts/contracts.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { EthMethod } from '../methods/methods';
+import { ContractReader } from '../../base/contracts/contractInstances';
+import { ensureExist } from '../../util/ensure';
+import { Contract, Instance } from '../../base/contracts/contracts';
+import { MethodDefinition, Methods, MethodsImpl } from '../../base/methods/methods';
+
+export class EthContract implements Contract<EthMethod> {
```
Comment (twtaylorbitgo) [inline src/eth/contracts/contracts.ts:7]: I like this a lot better. these are named well!

File: src/eth/decoder/types.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  } else {
+    switch (parsePrimitive(type)) {
+      case Primitive.Address:
+        return addHexPrefix(value);
+      case Primitive.Bool:
+        return formatBool(value);
+      case Primitive.Bytes:
+        return bufferToHex(value);
+      case Primitive.Int:
+        // value is a Buffer
```
Comment (twtaylorbitgo) [inline src/eth/decoder/types.ts:27]: let's type check this then and assert it

**Review by twtaylorbitgo — CHANGES_REQUESTED** [review]
* let's try to get the base generics in interface better names
* consistency with _ vs no _
* overall structure looks good, liking what I'm seeing

**Review by twtaylorbitgo — CHANGES_REQUESTED** [review]
hmmm let's also try to get some test coverage for these base classes

File: src/base/contracts/contractInstances.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  private parseMethods(definitions: MethodDefinition[]): M[] {
+    return definitions.reduce((collector: M[], def: MethodDefinition) => {
+      const method: M = new this._methodClass(def);
+      collector.push(method);
+      return collector;
+    }, []);
+  }
+}
+
+export interface ContractInstances {
```
[the remark this answers is not in the document]
Comment (FabianSimtlix) [reply src/base/contracts/contractInstances.ts:95]: Actually I think it should be `ContractInstance`

File: src/eth/decoder/types.ts (hunk 3 above)
Comment (ipelaez-simtlix) [reply src/eth/decoder/types.ts:27]: There are type check for this here : https://github.com/BitGo/eth-contracts/blob/STLX-5_generic/src/base/decoder/types.ts#L40

Comment (FabianSimtlix) [reply src/eth/decoder/types.ts:27]: Is already type checked at the parser

File: src/base/methods/methods.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  id: string;
+  name: string;
+  call(params: { [key: string]: any }): any;
+  explain(): MethodDefinition;
+}
+
+/**
+ * This represents the set of methods that conform a contract instance
+ */
+export interface Methods<M extends Method> {
```
[the remark this answers is not in the document]
Comment (FabianSimtlix) [reply src/base/methods/methods.ts:28]: This represents a collection of methods that's why is pluralized

**Comment by twtaylorbitgo** [comment]
> hmmm let's also try to get some test coverage for these base classes

disregard - isma was able to provide good coverage!

File: src/base/contracts/contractInstances.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+ * These are stored locally as JSON ABI definition files
+ */
+export function listContractTypes(abiDirPath: string): string[] {
+  return fs.readdirSync(path.join( __dirname, abiDirPath)).map((fileName: string) => {
+    ensure(fileName.endsWith('.json'), `Malformed JSON abi filename: ${fileName}`);
+    return fileName.replace('.json', '');
+  });
+}
+
+export class ContractReader<M extends Method, T extends Methods<M>> {
```
[the remark this answers is not in the document]
Comment (fmeichtri-simtlix) [reply src/base/contracts/contractInstances.ts]: Thanks! The names for the generics were fixed. Please check them

File: src/base/contracts/contractInstances.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  protected readonly abiDirPath: string;
+  protected readonly configDirPath: string;
+  protected readonly defaultInstanceKey: string;
+  readonly _methodsClass: MethodsClass<M, T>;
+  readonly _methodClass: MethodClass<M>;
```
[the remark this answers is not in the document]
Comment (fmeichtri-simtlix) [reply src/base/contracts/contractInstances.ts]: Done

File: src/base/contracts/contractInstances.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (fmeichtri-simtlix) [reply src/base/contracts/contractInstances.ts]: Done. It name was changed to  `ContractInstance`

File: src/base/decoder/types.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    Int,
+}
+
+export function formatBool(value: Buffer): boolean {
+  const intBool = bufferToInt(value);
+  return intBool === 1;
+}
+
+export function isArray (type: string): boolean {
+  return type.lastIndexOf(']') === type.length - 1;
```
[the remark this answers is not in the document]
Comment (fmeichtri-simtlix) [reply src/base/decoder/types.ts]: Here I added the following validation: 
`return type.includes('[]') && type.lastIndexOf(']') === type.length - 1;`
the parameters given to the function are for example: `uint256[]` `address[]`

File: src/base/methods/methods.ts (hunk 1 above)
Comment (fmeichtri-simtlix) [reply src/base/methods/methods.ts]: Fixed, Thanks Thomas

File: src/eth/decoder/types.ts (hunk 3 above)
Comment (twtaylorbitgo) [reply src/eth/decoder/types.ts:27]: ok good

**Review by twtaylorbitgo — APPROVED** [review]

**Pull request merged**